shard/connpool.lua

```diff
@@ -221,34 +252,33 @@ end
 
 -- heartbeat worker
 local function heartbeat_fiber(self)
-    fiber.name("heartbeat")
-    local i = 0
+    fiber.name("_" .. self.configuration.pool_name .. "_heartbeat")
     while true do
-        i = i + 1
         -- random select node to check
         local server = self:one(nil, true)
 
         if server ~= nil then
             local uri = server.uri
             log.debug("checking %s", uri)
 
-            if server.conn == nil then
-                log.warn('set -1 to all')
+            if server.conn ~= nil and server.conn.state ~= 'active' then
+                -- failed server's opinion marked useless
                 for _, opinion in pairs(self.heartbeat_state[server.uri]) do
                     opinion.ts = fiber.time()
                     opinion.try = INFINITY_MIN
                 end
-
+                -- register failed attempt at heartbeat
                 if self.self_server then
-                    self.heartbeat_state[self.self_server.uri][server.uri] = {
-                        ts = fiber.time(), try = INFINITY_MIN}
+                    local opinion = self.heartbeat_state[self.self_server.uri][server.uri]
+                    opinion.ts = fiber.time()
+                    opinion.try = opinion.try + 1
                 end
             else
                 -- get heartbeat from node
                 local response
                 local status, err_state = pcall(function()
-                        local expr = "return heartbeat('" .. self.configuration.pool_name .. "')"
-                        response = server.conn:timeout(self.HEARTBEAT_TIMEOUT):eval(expr)
+                    local expr = "return heartbeat('" .. self.configuration.pool_name .. "')"
+                    response = server.conn:timeout(self.HEARTBEAT_TIMEOUT):eval(expr)
```

Comment:
'Timeout' function may be called on nil, because 'if' statement trigger if server.conn == nil.

Reply:
it is under `pcall`, so such situation is handled accordingly